### PR TITLE
fix: use existing query params from blobstore response 

### LINKED ISSUE #132: Unable to push blobs to docker registry
### What happened in your code base?

Hi, 

I've attempted to upload a blob using the following code however it fails with the following error

Error response from registry: {"errors":[{"code":"BLOB_UPLOAD_INVALID","message":"blob upload invalid","detail":{}}]}

Registry Log Error: 2024-08-07 21:18:46 time="2024-08-07T20:18:46.952219413Z" level=error msg="response completed with error" err.code="blob upload invalid" err.detail="invalid secret" err.message="blob upload invalid" go.version=go1.20.8 http.request.contenttype="application/octet-stream" http.request.host=localhost http.request.id=dc571b23-ea56-4b20-b670-1874999be808 http.request.method=PUT http.request.remoteaddr="172.17.0.1:57354" http.request.uri="/v2/installationbundle/blobs/uploads/45ce63a0-a429-4185-b5ac-23874cb9d2c9?digest=sha256%3a44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a" http.request.useragent= http.response.contenttype="application/json; charset=utf-8" http.response.duration=1.160542ms http.response.status=404 http.response.written=88 vars.name=installationbundle vars.uuid=45ce63a0-a429-4185-b5ac-23874cb9d2c9 
2024-08-07 21:18:46 172.17.0.1 - - [07/Aug/2024:20:18:46 +0000] "PUT /v2/installationbundle/blobs/uploads/45ce63a0-a429-4185-b5ac-23874cb9d2c9?digest=sha256%3a44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a HTTP/1.1" 404 88 "" ""

I believe the issue is due to the fact that the _state query parameter from the registry response is being discarded with the current PushAsync method.  I modified the code below and after this I was able to push blobs into the registry.

```
    public async Task PushAsync(Descriptor expected, Stream content, CancellationToken cancellationToken = default)
    {
        var url = new UriFactory(Repository.Options).BuildRepositoryBlobUpload();
        using (var response = await Repository.Options.HttpClient.PostAsync(url, null, cancellationToken).ConfigureAwait(false))
        {
            if (response.StatusCode != HttpStatusCode.Accepted)
            {
                throw await response.ParseErrorResponseAsync(cancellationToken).ConfigureAwait(false);
            }

            var location = response.Headers.Location ?? throw new HttpRequestException("missing location header");
            url = location.IsAbsoluteUri ? location : new Uri(url, location);
        }

        // monolithic upload
        // add digest key to query string with expected digest value
        var req = new HttpRequestMessage(HttpMethod.Put, new UriBuilder(url)
        {
            //Query = $"digest={HttpUtility.UrlEncode(expected.Digest)}"
            Query = url.Query + "&" + $"digest={HttpUtility.UrlEncode(expected.Digest)}"
        }.Uri);
        req.Content = new StreamContent(content);
        req.Content.Headers.ContentLength = expected.Size;
```

### What did you expect to happen?

Blobs can be pushed

### How can we reproduce it?

My example code
```
var repository = await _registry.GetRepositoryAsync(repo, CancellationToken.None);

var emptyJsonContent = "{}";
var emptyJsonBytes = Encoding.UTF8.GetBytes(emptyJsonContent);
var emptyJsonDescriptor = new Descriptor
{
    MediaType = "application/vnd.unknown.config.v1+json",
    Size = emptyJsonBytes.Length,
    Digest = "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a"
};

using (var emptyJsonStream = new MemoryStream(emptyJsonBytes))
{
    await repository.Blobs.PushAsync(emptyJsonDescriptor, emptyJsonStream, CancellationToken.None);
}
```

### What is the version or commit of the ORAS .NET library?

Commit: c58adc96ac2d31cbd98d4c2e5211981326151ee4

### What are your OS and Runtime environments?

Windows 11 .net 8

### Are you willing to submit PRs to fix it?

- [X] Yes, I am willing to fix it.

### PR DESCRIPTION
**Pull request by daniel-pebble**
### What this PR does / why we need it
This resolves an issue when pushing to a docker registry, it responses with a _state query param, this must be part of the put request.  This PR, takes the query params from the response and appends the digest to the put request.

### Which issue(s) this PR resolves / fixes
Resolves / Fixes #132 

### Please check the following list
- [ ] Does the affected code have corresponding tests, e.g. unit test, E2E test? yes
- [ ] Does this change require a documentation update? No
- [ ] Does this introduce breaking changes that would require an announcement or bumping the major version? no
- [ ] Do all new files have an appropriate license header? yes

### CODE REVIEW

**Comment by shizhMSFT** [comment]
@daniel-pebble Thank you for your contribution. This PR contains the change of #135. Could you keep cc915c6561bd9787fdc4192046350683ddea6773 in this PR?

**Comment by daniel-pebble** [comment]
Hi @shizhMSFT, i've closed PR #135 and will commit under this PR to keep things simple.  I've added some unit tests for the validate digest function.  I'm not sure how I would go about adding unit tests for _camelCase naming convention, I've used sonarqube in the past to enforcing code styling, perhaps this is something Codecov can do?  Happy to revert these changes if this requires a wider discussion?

**Comment by daniel-pebble** [comment]
The other option of course is to use .editorconfig which would enforce _camalCase for private fields and PascalCase for public members.

I would suggest the following,  Whats your thoughts on this?

# Naming Styles
dotnet_naming_rule.private_fields_should_start_with_underscore.severity = error
dotnet_naming_rule.private_fields_should_start_with_underscore.symbols = private_fields
dotnet_naming_rule.private_fields_should_start_with_underscore.style = private_underscore_style

dotnet_naming_symbols.private_fields.applicable_kinds = field
dotnet_naming_symbols.private_fields.applicable_accessibilities = private
dotnet_naming_symbols.private_fields.required_modifiers = 

dotnet_naming_style.private_underscore_style.capitalization = camel_case
dotnet_naming_style.private_underscore_style.required_prefix = _

dotnet_naming_rule.public_members_should_start_with_uppercase.severity = error
dotnet_naming_rule.public_members_should_start_with_uppercase.symbols = public_members
dotnet_naming_rule.public_members_should_start_with_uppercase.style = public_uppercase_style

dotnet_naming_symbols.public_members.applicable_kinds = property, field, method
dotnet_naming_symbols.public_members.applicable_accessibilities = public
dotnet_naming_symbols.public_members.required_modifiers = 

dotnet_naming_style.public_uppercase_style.capitalization = pascal_case

**Comment by shizhMSFT** [comment]
`.editorconfig` may work and worth a try.

File: src/OrasProject.Oras/Content/Digest.cs [hunk 1]
```diff
@@ -13,26 +13,37 @@
...
+        if (string.IsNullOrEmpty(digest) || !_digestRegex.IsMatch(digest))
         {
             throw new InvalidDigestException($"Invalid digest: {digest}");
         }
+
+        var algorithm = digest.Split(':')[0];
+        if (!_supportedAlgorithms.Contains(algorithm))
+        {
+            throw new InvalidDigestException($"Unrecognized, unregistered or unsupported digest algorithm: {algorithm}");
+        }
```
Comment (shizhMSFT) [inline src/OrasProject.Oras/Content/Digest.cs:45]: We indeed need a better `Digest` implementation in the future as currently `oras-dotnet` only supports `sha256`.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**